### ISSUE TITLE
[Q] Is it acceptable to supply a macos port with the release?

### ISSUE THREAD

**Issue by ivan-mogilko** [adventuregamestudio/ags#1333]
This was bothering me for a while, I know that CI is building mac port binary, but we don't have it uploaded to the github releases, neither posted in release thread on forums etc. That probably makes it less likely for users to know of its existence.

I realize that normally a mac app should be packages in a certain way to become a legitimate osx app.

But still may there be a reason to have an engine binary for things like quick introduction, random game testing, and so on?

**Comment by ericoporto** [comment]
the steps that are done like `mygame.app` and the `info.plist` could be generated by the editor, but the codesigning and the notarization which ARE needed to release for macOS afaik requires a MacOS computer to do the steps.

On packaging, the best information is here : http://www.edenwaith.com/blog/index.php?p=112 , so up to until the paragraph that reads `Code Signing` I believe is possible to do in the Editor.

**Comment by hobbesjaap** [comment]
It's my understanding that if I create the app, I can run it on my Mac without notarization & codesigning. To release wider, these two things do need to happen. Unless we want people to have to run xattr -cr commands on unsigned apps every time they download them.

So for an unsigned build, this would work I think. I've currently got the "compile for Mac" thing done as posted on the forum the other day. I've now symbolic-linked my "Data" folder (when I compile my game in the editor) to the "Resources" folder in the Mac game wrapper. So whatever I do, I get a ready-to-go Mac-native app of my game's latest development version. A bit cumbersome, so if AGS could spit out a native Mac app on its own? I can see value in that. Lots of freeware stuff released on the internet is unsigned, with the understanding for a Mac user to right-click & open instead. The xattr-cr is really something that seems to be required lately with Big Sur & M1 coming in.

**Comment by ericoporto** [comment]
First question to answer to supplying apple binaries for me is how does backward compatibility works. If I build something on Big Sur, does it run on High Sierra? What's the expectations on this matter? Because we would be building binaries always on the latest version because of CI limitations.

**Comment by hobbesjaap** [comment]
True, I think quite a few apps extend back quite far in compatibility mode, it's more the future-proofing aspect of things that I was thinking about. As I mentioned on the forum, a few things I feel the MacOS port will need moving forward:

- Support for Metal (either native or through something like MoltenVK): Metal has been integrated into MacOS since El Capitan (10.11) which is about 5 years ago. OpenGL has been deprecated for the past 2 to 3 years, by my very rough estimation so the writing's a bit on the wall with that one, I fear. I hope they don't drop it, but the OpenGL version on MacOS is severely limited & outdated compared to Windows/Linux, for example. Apple's very niche in the gaming world so I also fully appreciate it's not much of a priority. 
- Support for the M1 chipset. I believe Apple have made the official statement to continue to support Intel Mac computers for the next 2 years. So again, the writing's on the wall. And with the continued synergy between MacOS and iOS/iPadOS/whateverOS, the move to a unified codebase is becoming more and more of a "big deal". As seen with MacOS 12 Monterey, which again makes further steps towards unified feature release between iOS & MacOS. So this support is either through a Universal Binary (seems like the smart move for now for backwards compatibility).

Now again, I'm well aware I'm making these comments in an almost flippant way since I know 0% about the work involved to make this happen. So the path to getting to these milestones can be very long indeed, but in order to futureproof the MacOS version of AGS, I think these two elements are key. So whilst the point of creating an M1 binary is not particularly relevant at this point, Apple is a company that's very happy to suddenly drop support for legacy hardware / software which makes it a bit of a pain for small-time open source projects like AGS. The upcoming move to SDL2 is already a great step towards futureproofing (for now) so perhaps there's opportunities to explore there?

I realise this thread is now at risk of veering significantly off-topic and raises potentially 2 feature requests. If desired, I could describe these in more detail in separate posts?

**Comment by ericoporto** [comment]
M1 -> it's just arm, we already have this.

Metal -> is already there through software renderer. A "proper" metal renderer would require someone not only writing one, but also maintaining it. So far no one has. I think a better strategy would be enhancing the SDL Renderer.

Better to split in separate topics.

**Comment by ericoporto** [comment]
An year later, our CI is running on Big Sur and Monterey, so we can now actually cross compile on it to Apple Silicon. So supposing a binary built on these newer MacOS systems can run on an older OS version, this could be feasible.

**Comment by ericoporto** [comment]
I just noticed that if someone unpacks it on Windows, set the files for their game and pack again it will be cursed by #471 too, because MacOS also requires ags binary to have executable flag set....

**Comment by ericoporto** [comment]
Hey, I will probably submit a PR at the weekend that does two things:

- build for arm64 and amd64 in a fat binary
- package the generated `.app` dir as a `.tar.gz` and push it as additional release artifact

I may additionally reconfigure the build so that release and debug reuse the same machine (to reduce the chances the vm is unavailable), and also use ninja instead of make. The release will be generated for "older" available image from cirrus-ci.

In the future if someone wants to add MacOS building in the editor, the way to do so would be the following:

- develop a program that takes the game files (probably in a zip) and a description file (probably xml to be easily editable in MacOS too), downloads the appropriate released MacOS package, and do the `.app` building with [signing and notarization as described here](https://www.adventuregamestudio.co.uk/forums/index.php?topic=47264.msg636643289#msg636643289)
- make so this program has a worker mode so that the editor can send files to it and retrieve the MacOS `.app` packaged in a convenient format that preserves execution flags (either `.dmg` or `.tar.gz`)

**Comment by ivan-mogilko** [comment]
Removed from 3.6.0 milestone; from what I understand few bigger changes have to be done first, so it's better to write task tickets based on this discussion and close this afterwards.

**Comment by ericoporto** [comment]
Minor update: I have no idea how to pick the oldest Xcode build that is the release build, and build JUST THAT on ags tag release, and also package JUST THAT in the packaging step of the CI. Paging @morganwillcock if you have any idea.

**Comment by morganwillcock** [comment]
I don't think it would be too much of a problem to select it, but I would worry what kind of guarantees are assumed or implied if the binary is provided with an AGS release.

1. It won't be signed or notarized
2. The version of Xcode used may prevent it running on older versions of macOS
3. The version of Xcode will be raised by the CI provider and we wouldn't have a way to remain on an older version

I can't see it is feasible to make AGS releases that include the binary and also maintain support of it because we may not have a way to rebuild it.

**Comment by ericoporto** [comment]
> The version of Xcode will be raised by the CI provider and we wouldn't have a way to remain on an older version

To be honest, Apple also raises it locally too, my Xcode doesn't even load if I don't update on Apple Store. The minimum current target right now on the latest is High Sierra. I believe everyone keeps up with Apple OS upgrades at least a bit. So, with latest tools, all below should be able to run a binary built with those 

- macOS Ventura | 13.3.1
- macOS Monterey | 12.6.4
- macOS Big Sur | 11.7.5
- macOS Catalina | 10.15.7
- macOS Mojave | 10.14.6
- macOS High Sierra | 10.13.6

I would say providing a release experimentally would allow to advance the knowledge on macOS compatibility.

My other approach is to yank out all of those extras in the matrix and just build the thing I actually care.

**Comment by morganwillcock** [comment]
> About signing and notarization it SHOULD NOT be signed or notarized, as we don't control the game that will be placed along.

From our perspective it should not be. From the perspective of someone downloading a game it should be. It is more a question of what providing the engine binary is implying in terms of continued support. The requirements to build the engine are a subset of the requirements needed to sign and notarize, so someone following Apple's guidelines already has everything they need. As an example, if macOS loses the ability to run unsigned binaries is that now an AGS bug or does that get ignored? If we are supplying binaries with no guarantees about where they run, does that also mean we should be including FreeBSD binaries with every release?

Do you want to just include the binaries in the GitHub release or would this get bundled with the Editor? Both are described in this issue. I think it would be fine to include it in the GitHub release and refer to it as experimental. I don't think it would be wise to try any type of integration with the Editor whilst the build process isn't reproducible. I did find these:
- https://github.com/kholia/OSX-KVM
- https://github.com/sickcodes/Docker-OSX

I'm sceptical on the legality, but maybe one of them provides a way to build that can work independently from the CI, without relying on one person's hardware.

**Comment by ivan-mogilko** [comment]
> Do you want to just include the binaries in the GitHub release or would this get bundled with the Editor? Both are described in this issue.

IIRC I opened this issue with the intent to do the first, as people we wondering on how to run on mac, and I was not able to give them any simple download instructions.

Deploying the game for mac from the Editor should be a separate task imo. Maybe it will require additional steps, and adding dedicated set of preferences to the editor or project settings.

**Comment by ericoporto** [comment]
> As an example, if macOS loses the ability to run unsigned binaries is that now an AGS bug or does that get ignored?

It can be considered that is "already lost" as is a significant effort to run something unsigned, but you can sign something unsigned, but you will sign the bundle. The bundle is the game the user is creating and we don't have control on that.

> If we are supplying binaries with no guarantees about where they run, does that also mean we should be including FreeBSD binaries with every release?

That is not the same case, in Xcode we can specify the minimum target and it will work, similarly on Android how we build with current tools and have limited compatibility with past Android versions.

In FreeBSD there is no such guarantee, it's more similar to how it works in Linux, and would need investigation on hoe whatever it's libc binary compatibility works.

> Do you want to just include the binaries in the GitHub release or would this get bundled with the Editor?

Only with the GitHub Release, they are useless on the Editor as one would need an actual macOS to finalize the process (sign and notarization). Plus, it's best people distribute what they actually test - and yes, I have been testing the binaries we produce, and they so far have worked for the past years.

> I'm sceptical on the legality, but maybe one of them provides a way to build that can work independently from the CI, without relying on one person's hardware.

Let's not do anything ilegal, our CI runs in actual Apple Silicon in actual macOS, lets use limit the scope to that.

**Comment by morganwillcock** [comment]
If you want to implement it, I think the best way would be to just use the hardcoded Xcode version number. The download is already available. When the CI provider changes the Xcode version, the CI failure will be a prompt to redo the release documentation to indicate which version of Xcode was used, and for someone to re-test it.

`https://api.cirrus-ci.com/v1/artifact/build/$CIRRUS_BUILD_ID/build_macos/binaries/bin_14.0_release/ags`

**Comment by ericoporto** [comment]
I can add this to the Editor, with the script for other things that HAS to be run on Mac for signing and notarization but you will still be able to run without and on macOS the person running a non notarized and non signed game will get the pop-up annoyances but it's still possible to run. 

Will people keep complaining about version and whatever of macOS is incorrect or can we just run in the actually supported by Apple versions of the OS? The way to go with Apple is just keep updating the OS and buying new hardware when that is not possible, and even SDL isn't keeping compatibility beyond so someone complaining will need to also maintain a fork of SDL.
